File: python/lib/requests.py
# Copyright 2015 ETH Zurich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
:mod:`requests` --- Request handling library
============================================
"""
# Stdlib
import logging
import threading
import queue
from collections import defaultdict

# External
from prometheus_client import Gauge

# SCION
from lib.thread import thread_safety_net
from lib.util import SCIONTime


# Exported metrics.
REQS_PENDING = Gauge("rh_requests_pending", "# of pending requests",
                     ["server_id", "isd_as", "type"])


class RequestHandler(object):
    """
    Small utility class to queue requests, check if they've been fulfilled, and
    send replies. Requests are expired after a certain period.

    This class is intended to be used in a seperate thread (run via
    RequestHandler.start()), and monitors a queue.Queue for notifications. To
    add a new request, the client code adds a key and request object to the
    queue. To signal that the RequestHandler should check for fulfilled queries
    for a given key, the client can simply send `None` instead of a request
    object.

    RequestHandler expects 4 parameters:
        - a queue to read from (discussed above)
        - a 'check' callback to see if a request can be fulfilled
        - a 'fetch' callback to send a query for the answer
        - a 'reply' callback to send the answer back to the requester
    """
    TTL = 2.0
    MAX_LEN = 16

    def __init__(self, queue, check, fetch, reply, ttl=TTL,
                 key_map=None, labels=None):  # pragma: no cover
        """
        :param queue.Queue queue:
            Used to receive request notifications, see class docstring for
            details.
        :param function check:
            Called with the provided key, should return `True` if the request
            can be answered, otherwise `False`.
        :param function fetch:
            Called with the provided key and request info, responsible for
            fetching the answer for the request.
        :param function reply:
            Called with the provided key and request info, should return the
            result to the requester.
        :param dict labels:
            Labels added to the exported metrics. The following labels are supported:
                - server_id: A unique identifier of the server that is exporting
                - isd_as: The ISD_AS of where the server is running
                - type: A generic label for the type of the revocations.
        """
        self._queue = queue
        self._check = check
        self._fetch = fetch
        self._reply = reply
        self._ttl = ttl
        self._req_map = defaultdict(list)
        self._key_map = key_map or self._def_key_map
        self._labels = labels
        if self._labels:
            self._init_metrics()

    def _init_metrics(self):  # pragma: no cover
        REQS_PENDING.labels(**self._labels).set(0)

    @classmethod
    def start(cls, name, *args, **kwargs):  # pragma: no cover
        """
        Create a queue and a RequestHandler object with the supplied
        args/kwargs, then start RequestHandler.run in a thread, and return the
        queue to the caller. This is the recommended way of running
        RequestHandler.
        """
        q = queue.Queue()
        inst = cls(q, *args, **kwargs)
        threading.Thread(
            target=thread_safety_net, args=(inst.run,),
            name=name, daemon=True).start()
        return q

    def run(self):
        while True:
            key, req = self._queue.get()
            if req:
                # Add a new request
                self._add_req(key, req)
            # Answer existing requests, if possible.
            for k in self._key_map(key, self._req_map.keys()):
                self._answer_reqs(k)

    def _add_req(self, key, request):
        self._expire_reqs(key)
        if not self._check(key):
            self._fetch(key, request)
        self._req_map[key].append((SCIONTime.get_time(), request))
        if self._labels:
            REQS_PENDING.labels(**self._labels).inc()

    def _answer_reqs(self, key):
        if not self._check(key):
            # Don't have the answer yet.
            return
        self._expire_reqs(key)
        reqs = self._req_map[key]
        count = 0
        while reqs:
            _, req = reqs.pop(0)
            self._reply(key, req)
            count += 1
        if self._labels:
            REQS_PENDING.labels(**self._labels).dec(count)
        del self._req_map[key]

    def _expire_reqs(self, key):
        if key not in self._req_map:
            return
        now = SCIONTime.get_time()
        count = 0
        for ts, req in self._req_map[key][:]:
            if now - ts > self._ttl:
                count += 1
                self._req_map[key].remove((ts, req))
        if count:
            logging.debug("Expired %d requests for %s", count, key)
            if self._labels:
                REQS_PENDING.labels(**self._labels).dec(count)

    @staticmethod
    def _def_key_map(key, keys):  # pragma: no cover
        if key in keys:
            return [key]
        return []

File: go/cs/metrics/requests.go
// Copyright 2019 Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package metrics

import (
	"github.com/prometheus/client_golang/prometheus"

	"github.com/scionproto/scion/go/lib/addr"
	"github.com/scionproto/scion/go/lib/ctrl/seg"
	"github.com/scionproto/scion/go/lib/infra/modules/segfetcher"
	"github.com/scionproto/scion/go/lib/prom"
	"github.com/scionproto/scion/go/proto"
)

// RequestOkLabels contains the labels for a request that succeeded.
type RequestOkLabels struct {
	SegType proto.PathSegType
	DstISD  addr.ISD
}

// Labels returns the labels.
func (l RequestOkLabels) Labels() []string {
	return []string{"seg_type", "dst_isd"}
}

// Values returns the values.
func (l RequestOkLabels) Values() []string {
	return []string{l.SegType.String(), l.DstISD.String()}
}

// RequestLabels contains the labels for requests.
type RequestLabels struct {
	Result string
	RequestOkLabels
}

// Labels returns the labels.
func (l RequestLabels) Labels() []string {
	return append([]string{"result"}, l.RequestOkLabels.Labels()...)
}

// Values returns the values.
func (l RequestLabels) Values() []string {
	return append([]string{l.Result}, l.RequestOkLabels.Values()...)
}

// WithResult returns the labels with the modified result.
func (l RequestLabels) WithResult(result string) RequestLabels {
	l.Result = result
	return l
}

// Request is for request metrics.
type Request struct {
	Requests        *prometheus.CounterVec
	SegmentsSent    *prometheus.CounterVec
	RevocationsSent *prometheus.CounterVec
}

func newRequests() Request {
	subsystem := "requests"
	return Request{
		Requests: prom.NewCounterVecWithLabels(PSNamespace, subsystem, "total",
			"Number of segment requests total. \"result\" indicates the outcome.",
			RequestLabels{}),
		SegmentsSent: prom.NewCounterVecWithLabels(PSNamespace, subsystem, "replied_segments_total",
			"Number of segments in reply to segment requests.", RequestOkLabels{}),
		RevocationsSent: prom.NewCounterVecWithLabels(
			PSNamespace,
			subsystem,
			"replied_revocations_total",
			"Number of revocations in reply to segments requests.",
			RequestOkLabels{},
		),
	}
}

// Count returns the counter for requests total.
func (r Request) Count(l RequestLabels) prometheus.Counter {
	return r.Requests.WithLabelValues(l.Values()...)
}

// RepliedSegs returns the counter for the number of segments in a seg reply.
func (r Request) RepliedSegs(l RequestOkLabels) prometheus.Counter {
	return r.SegmentsSent.WithLabelValues(l.Values()...)
}

// RepliedRevs returns the counter for the number of revocations in a seg
// reply.
func (r Request) RepliedRevs(l RequestOkLabels) prometheus.Counter {
	return r.RevocationsSent.WithLabelValues(l.Values()...)
}

// DetermineReplyType determines which type of segments is in the reply. The
// method assumes that segs only contains one type of segments.
func DetermineReplyType(segs segfetcher.Segments) seg.Type {
	if len(segs) > 0 {
		return segs[0].Type
	}
	return 0
}
